File: tdoa_Collection.py
import uhd
import numpy as np
import time

# 参数设置
center_freq = 2.4e9  # 2.4 GHz
sample_rate = 56e6  # 0.5 MSps
gain = 40  # 接收增益
num_samples = int(0.00001* sample_rate)  # 采集 2 秒的数据
buffer_size =32768 # 缓冲区大小
file_name = "signal1.bin"

# 初始化 USRP
usrp = uhd.usrp.MultiUSRP()
usrp.set_rx_rate(sample_rate)
usrp.set_rx_freq(uhd.libpyuhd.types.tune_request(center_freq))
usrp.set_rx_gain(gain)
usrp.set_clock_source("external", 0)  # 使用 GPSDO 时钟
usrp.set_time_source("external", 0)  # 使用 GPSDO 时间源
print(f"当前 USRP 时间源: {usrp.get_time_source(0)}")
print(f"当前 USRP 时钟源: {usrp.get_clock_source(0)}")
usrp.set_time_now(uhd.types.TimeSpec(0.0))
time.sleep(1)  # 等待 1 秒，确保 USRP 处理命令

# 确保 USRP 已检测到 GPSDO 的 PPS 信号
last_pps_time = usrp.get_time_last_pps().get_real_secs()
print(f"当前 USRP 时间: {usrp.get_time_now().get_real_secs()}")
print(f"最近的 PPS 触发时间: {last_pps_time}")

# 等待下一个 PPS 触发
while True:
    new_pps_time = usrp.get_time_last_pps().get_real_secs()
    if new_pps_time > last_pps_time:
        break
    time.sleep(0.001)

print(f"PPS 触发，当前 USRP 时间: {usrp.get_time_now().get_real_secs()}")

# 在 PPS 触发前立即设置下一个 PPS 触发时刻
usrp.set_time_next_pps(uhd.types.TimeSpec(9.0))

# 再次等待 PPS 触发
last_pps_time = usrp.get_time_last_pps().get_real_secs()
while True:
    new_pps_time = usrp.get_time_last_pps().get_real_secs()
    if new_pps_time > last_pps_time:
        break
    time.sleep(0.001)

# 记录 PPS 触发时间
pps_time = usrp.get_time_now().get_real_secs()
print(f"第二次 PPS 触发，当前 USRP 时间: {pps_time}")

# 创建流
stream_args = uhd.usrp.StreamArgs("fc32", "sc16")
stream_args.args = "recv_buff_size=10000000,recv_frame_size=65536,num_recv_frames=512"
streamer = usrp.get_rx_stream(stream_args)
metadata = uhd.types.RXMetadata()
recv_buffer = np.zeros((buffer_size,), dtype=np.complex64)
received_samples = np.zeros((num_samples,), dtype=np.complex64)  # 预分配数组

# 等待 1 秒以确保时间稳定
time.sleep(0) 

# 启动接收流
streamer.issue_stream_cmd(uhd.types.StreamCMD(uhd.types.StreamMode.start_cont))
start_time = usrp.get_time_now().get_real_secs()
print(f"信号采集开始，当前 USRP 时间: {start_time}")

# 记录信号
write_index = 0
while write_index < num_samples:
    num_recv = streamer.recv(recv_buffer, metadata, timeout=1.0)
    
    if metadata.error_code == uhd.types.RXMetadataErrorCode.overflow:
        print("RX Metadata warning: Overflow detected, continuing...")
        continue  # 忽略溢出错误，继续采集
    elif metadata.error_code != uhd.types.RXMetadataErrorCode.none:
        print(f"RX Metadata error: {metadata.error_code}")
        break  # 其他错误则终止

    # 确保不超过数组边界
    num_samples_to_copy = min(num_recv, num_samples - write_index)
    
    # 直接存入 NumPy 数组
    received_samples[write_index : write_index + num_samples_to_copy] = recv_buffer[:num_samples_to_copy]
    write_index += num_samples_to_copy

# 停止流
streamer.issue_stream_cmd(uhd.types.StreamCMD(uhd.types.StreamMode.stop_cont))

end_time = usrp.get_time_now().get_real_secs()
print(f"信号采集完成，当前 USRP 时间: {end_time}")

# 保存数据
received_samples.tofile(file_name)

print(f"信号采集完成，已保存到 {file_name}，PPS 时间戳: {pps_time}，信号采集时间戳: {start_time}")

File: tdoa_Collection.cpp
#include <uhd/usrp/multi_usrp.hpp>
#include <uhd/stream.hpp>
#include <uhd/types/tune_request.hpp>
#include <uhd/types/metadata.hpp>
#include <iostream>
#include <fstream>
#include <thread>
#include <chrono>
#include <complex>
#include <vector>
#include <cstring>

int main() {
    try {
        // 参数设置
        const double center_freq = 2.4e9;
        const double sample_rate = 15e6;
        const double gain = 40;
        const size_t num_samples = static_cast<size_t>(0.005 * sample_rate);
        const size_t buffer_size = 32768;
        const std::string file_name = "signal1.bin";

        // 创建 USRP
        uhd::device_addr_t dev_addr;
        auto usrp = uhd::usrp::multi_usrp::make(dev_addr);

        usrp->set_rx_rate(sample_rate);
        usrp->set_rx_freq(uhd::tune_request_t(center_freq));
        usrp->set_rx_gain(gain);

        usrp->set_clock_source("external", 0);
        usrp->set_time_source("external", 0);
        std::cout << "当前时钟源: " << usrp->get_clock_source(0) << std::endl;
        std::cout << "当前时间源: " << usrp->get_time_source(0) << std::endl;

        usrp->set_time_now(uhd::time_spec_t(0.0));
        std::this_thread::sleep_for(std::chrono::seconds(1));

        // 确保 GPSDO PPS 锁定
        double last_pps = usrp->get_time_last_pps().get_real_secs();

        // 等待 PPS 上升沿
        while (true) {
            double new_pps = usrp->get_time_last_pps().get_real_secs();
            if (new_pps > last_pps) break;
            std::this_thread::sleep_for(std::chrono::milliseconds(1));
        }

        // 设置下一个 PPS 触发时间
        usrp->set_time_next_pps(uhd::time_spec_t(9.0));

        last_pps = usrp->get_time_last_pps().get_real_secs();
        while (true) {
            double new_pps = usrp->get_time_last_pps().get_real_secs();
            if (new_pps > last_pps) break;
            std::this_thread::sleep_for(std::chrono::milliseconds(1));
        }

        // 配置数据流
        uhd::stream_args_t stream_args("fc32", "sc16");
        stream_args.args["recv_buff_size"] = "10000000";
        stream_args.args["recv_frame_size"] = "65536";
        stream_args.args["num_recv_frames"] = "512";

        auto rx_stream = usrp->get_rx_stream(stream_args);
        uhd::rx_metadata_t md;
        std::vector<std::complex<float>> recv_buffer(buffer_size);
        std::vector<std::complex<float>> received_samples(num_samples);

        std::this_thread::sleep_for(std::chrono::milliseconds(500));

        // 启动流
        uhd::stream_cmd_t stream_cmd(uhd::stream_cmd_t::STREAM_MODE_START_CONTINUOUS);
        stream_cmd.stream_now = true;
        rx_stream->issue_stream_cmd(stream_cmd);

        auto ts = usrp->get_time_now();
        uint64_t start_sec = ts.get_full_secs();
        uint64_t start_nsec = static_cast<uint64_t>(ts.get_frac_secs() * 1e9);
        std::cout << "开始接收，USRP 时间: " << start_sec << " 秒 " << start_nsec << " 纳秒" << std::endl;

        size_t write_index = 0;
        while (write_index < num_samples) {
            size_t num_rx = rx_stream->recv(&recv_buffer.front(), buffer_size, md, 1.0);

            if (md.error_code == uhd::rx_metadata_t::ERROR_CODE_OVERFLOW) {
                std::cerr << "警告: 溢出发生，继续接收..." << std::endl;
                continue;
            } else if (md.error_code != uhd::rx_metadata_t::ERROR_CODE_NONE) {
                std::cerr << "接收错误: " << md.strerror() << std::endl;
                break;
            }

            size_t to_copy = std::min(num_rx, num_samples - write_index);
            std::memcpy(&received_samples[write_index], &recv_buffer[0], to_copy * sizeof(std::complex<float>));
            write_index += to_copy;
        }

        // 停止流
        uhd::stream_cmd_t stop_cmd(uhd::stream_cmd_t::STREAM_MODE_STOP_CONTINUOUS);
        rx_stream->issue_stream_cmd(stop_cmd);

        auto ts_end = usrp->get_time_now();
        uint64_t end_sec = ts_end.get_full_secs();
        uint64_t end_nsec = static_cast<uint64_t>(ts_end.get_frac_secs() * 1e9);
        std::cout << "接收完成，USRP 时间: " << end_sec << " 秒 " << end_nsec << " 纳秒" << std::endl;

        // 写入文件
        std::ofstream outfile(file_name, std::ios::binary);
        outfile.write(reinterpret_cast<const char*>(received_samples.data()), received_samples.size() * sizeof(std::complex<float>));
        outfile.close();

        std::cout << "已保存信号到 " << file_name << "\n采集起始时间: " << start_sec << " 秒 " << start_nsec << " 纳秒" << std::endl;

    } catch (const std::exception& e) {
        std::cerr << "发生异常: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
